### ISSUE TITLE
Mason: Spack Integration

### ISSUE THREAD

**Issue by Spartee** [chapel-lang/chapel#10488]
With this integration, Chapel users will not only be able to use dependencies on their machines through the `mason system` command, but also be able to download, install, search for, and find packages that are not on their system through the `mason external` command. Chapel users will use the `mason external` command to interface with a number of options to use external dependencies that closely resemble the existing mason commands.

The first external package manager Mason will be interfacing with is Spack. For more information on spack see the [documenation](https://spack.readthedocs.io/en/latest/) or the issue about [external dependencies](https://github.com/chapel-lang/chapel/issues/10037)

### Features (check = working in branch not master)
- [x] Install and setup Spack
- [x] Install Package
    - In Spack spec grammer form or not. e.g. (openssl@0.1.0%gcc or openssl)
- [x] Uninstall package
- [x] List installed packages
- [x] Search for packages
- [x] Show info for a package
- [x] List available compilers
- [x] Find installed compilers
- [x] Print architecture information
- [x] Help options
   - Search help
   - Install/uninstall help
   - Compiler help
   - Spack spec help

A `Mason.toml` for an external package that requires `zlib`, `openssl`, and `openblas` would look as follows:

`Mason.toml`

```toml
[brick]
name = "package"
version = "0.1.0"
chplVersion = "1.18.0"
author = "Sam Partee"

[dependencies]

[external]
openssl = "1.0.2k"
openblas = "0.2.20"
zlib = "1.2.11"

```

The build information included in the lock file would look as follows:

`Mason.lock`

``` toml
[root]
name = "package"
version = "0.1.0"
chplVersion = "1.18.0..1.18.0"
author = "Sam Partee"

[external]

[external.openblas]
name = "openblas"
version = "0.2.20"
libs = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/gcc-8.1.0/openblas-0.2.20-btofrthi6uvbkr6nxihksesh337spbn6/lib"
include = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/gcc-8.1.0/openblas-0.2.20-btofrthi6uvbkr6nxihksesh337spbn6/include"

[external.openssl]
name = "openssl"
version = "1.0.2k"
libs = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/openssl-1.0.2k-fwnsee6qcvbbgvmgp3f5hio6dwg6nh2d/lib"
include = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/openssl-1.0.2k-fwnsee6qcvbbgvmgp3f5hio6dwg6nh2d/include"

[external.zlib]
name = "zlib"
version = "1.2.11"
libs = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/zlib-1.2.11-cpdvq4e7otjepbwdtxmgk5bzszze27fj/lib"
include = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/zlib-1.2.11-cpdvq4e7otjepbwdtxmgk5bzszze27fj/include"
```

The build information is gathered through the `spack find -p <package>` command and parsed to return the path to the include and library directory.

This build information is then fed to `mason build` through compilation options just like how they would be specified on the command line. This process mimics the process of `Mason System`(pkg-config integration #10289).

External and System packages can also be used in conjunction as the `Mason.toml` below displays. This Mason.toml is for an example of the `LinearAlgebra` library that requires `openBLAS` and `LAPACK`. 

`Mason.toml`

```toml
[brick]
name = "cholesky"
version = "0.1.0"
chplVersion = "1.18.0"
compopts = "--ccflags -Wno-enum-conversion --ccflags -Wno-strict-prototypes -lopenblas"

[dependencies]

[external]
openblas = "0.2.20"

[system]
lapack = "*"
```

The build information for the `Mason.toml would look as follows:

`Mason.lock`

```toml
[root]
name = "cholesky"
compopts = "--ccflags -Wno-enum-conversion --ccflags -Wno-strict-prototypes -lopenblas"
version = "0.1.0"
chplVersion = "1.18.0..1.18.0"

[external]

[external.openblas]
name = "openblas"
version = "0.2.20"
libs = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/gcc-8.1.0/openblas-0.2.20-btofrthi6uvbkr6nxihksesh337spbn6/lib"
include = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/gcc-8.1.0/openblas-0.2.20-btofrthi6uvbkr6nxihksesh337spbn6/include"

[system]

[system.lapack]
name = "lapack"
version = "3.8.0"
libs = "-L/usr/local/Cellar/lapack/3.8.0_1/lib -llapack"
include = "/usr/local/Cellar/lapack/3.8.0_1/include"

```

**Comment by ben-albrecht** [comment]
@mppf - I believe you expressed some interest in the Spack integration design

**Comment by mppf** [comment]
@Spartee - cool, thanks.

If an external package depends on other external (spack) packages, will the include/libs in the .lock file just turn into arrays to have multiple elements?

The Spack design pays careful attention to variant configurations of packages (e.g. threaded vs not-threaded MPI). How can a mason user specify such things? I'd imagine e.g. advanced users could provide a Spack spec string as a dependency. E.g. see http://spack.readthedocs.io/en/latest/basic_usage.html#specs-dependencies for a complex example.

I'm very curious about if we can teach Spack about Mason packages too. Have you learned anything about that?

**Comment by Spartee** [comment]
@mppf Here is a more complex example that a recently cooked up with the HDF5 module. In short, I made the module a mason package. Below will be the description of how I did so. Hopefully this highlights some design points and helps drive our discussions.

## HDF5 as a mason package

1) `mason new masonHDF5`: create the new mason directory.
2) I moved the HDF5 module into `masonHDF5/src` , two tests for reading data into `masonHDF5/test`, and the `.h5` datasets into a new directory `masonHDF5/data`
3) I made minor changes to the tests and module so that the names and locations of the files matched up.
4) `mason external install hdf5@1.10.1%clang~mpi+hl`: to install the hdf5 module without mpi(dont need that for testing) and with the high level interface. 
5) `mason add --external hdf5@1.10.1%clang`: to add the dependency to my `Mason.toml` which looked like this after(see design questions below): 


```toml 
[brick]
name = "masonHDF5"
version = "0.1.0"
chplVersion = "1.18.0"
compopts = "-lhdf5 -lhdf5_hl"

[dependencies]

# must give name, version, compiler
# TODO: allow more flexibility in spec listing
[external]
hdf5 = "hdf5@1.10.1%clang" 
```
6) `mason update`: to lock in the build info in `Mason.lock` which looked like:

```toml

[root]
name = "masonHDF5"
compopts = "-lhdf5 -lhdf5_hl"
version = "0.1.0"
chplVersion = "1.18.0..1.18.0"

[external]

[external.hdf5]
name = "hdf5"
compiler = "clang"
dependencies = ["zlib"]
version = "1.10.1"
libs = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/hdf5-1.10.1-5aapkeur4ftry73anvfq5lr4dbcamzhb/lib"
include = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/hdf5-1.10.1-5aapkeur4ftry73anvfq5lr4dbcamzhb/include"

[external.hdf5.zlib]
name = "zlib"
compiler = "clang"
version = "1.2.8"
libs = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/zlib-1.2.8-iujafzqgr2zrpnerb7atokhazujeakns/lib"
include = "/Users/spartee/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/zlib-1.2.8-iujafzqgr2zrpnerb7atokhazujeakns/include"

```
7) `Mason test`: Which gathered the `masonHDF5` tests, compiled them, and then ran them with the following output from `ex_lite2.chpl`:

``` text
Updating mason-registry for mason-registry
compiled ex_lite2.chpl successfully
  1  2  3
  4  5  6

--------------------

--- Results ---
Test: ex_lite2 Passed

--- Summary:  1 tests run ---
-----> 1 Passed
-----> 0 Failed
```

As you can see the build info included is only just enough to build the package correctly. We might want to expand on that further. In the case above, `zlib`, is the dependency for `HDF5`, so it gets sub-tabled under the `[external.hdf5]` table. 

## Open design questions:
#### How do we want the mason.toml to look? as in:
```toml
[external]
hdf5 = "hdf5@1.10.1%clang" 
```
or 
```toml
[external]
hdf5 = "1.10.1%clang" 
```

#### How do we want to handle resolution?
- It would be pretty easy to slap our IVRS in here and make it do what it does best..complain..
- I am not sure of a way to make it so that Spack does this resolution for us at the moment but that could be an option given some instruction

#### What other commands do we want?
- some that I have been thinking about are
  - `dir`: cd to project dir to look at include, lib.. etc..
  - `info`: print info about the project
  - `spec-help`: prints spack's spec help
  - others...?

Thanks for reading!

**Comment by mppf** [comment]
> # must give name, version, compiler
> # TODO: allow more flexibility in spec listing

Can you say more about why these are all presently required?


```
[external]
hdf5 = "hdf5@1.10.1%clang" 
```

vs

```
[external]
hdf5 = "1.10.1%clang" 
```

I think it's appealing not to break up the Spack spec string (treating it as totally opaque has advantages) but the 1st version above also looks redundant.

What about
```
[external]
spack ="hdf5@1.10.1%clang" 
```

or even

```
[external]
"hdf5@1.10.1%clang" 
```

?

**Comment by Spartee** [comment]
All three fields(package, version, compiler) are required as they provide adequate specification as to which package should be pulled. Any less and there could be collisions between them. (which there still can be, but only collisions between variants aka.. with mpi/without mpi). 

```
[external]
spack ="hdf5@1.10.1%clang" 
```

This option presents an interesting possibility of interfacing with more than just spack in this table, but resolution across package managers is not something we want to tackle right now. Otherwise I am not super thrilled about how it looks. I prefer the redundant version honestly, because it prevents users from being able to list two of the same package in a `Mason.toml`(can't have two of the same key). 

I don't think we should split up the spec either. I have setup the external integration to pass specs around on purpose. There are a large number of variants that can accompany some of these packages, and trying to decipher between which is which without a spec is difficult. 


Also, this is not valid toml 
```
[external]
"hdf5@1.10.1%clang" 
```

so..


I think we go with
```
[external]
hdf5 = "hdf5@1.10.1%clang"
```
but to stay consistent we also put in support to make it look like our current format
```
[external]
hdf5 = "1.10.1"
```
For this one we will have to guess which package they want if they have more than one instance of the package with the same version built with two different compilers. We could probably make some inferences based on architecture.

**Comment by mppf** [comment]
@Spartee - Regarding the compilers, wouldn't we normally want to default to the same C compiler currently used with Chapel?

As far as converting
``` chapel
[external]
hdf5 = "1.10.1%clang"
```

into 

```
"1.10.1%clang"
```

if we require the version number to be first, isn't it just a matter of adding a `@` between the key and the value?

**Comment by Spartee** [comment]
@mppf Yes I think we will in the case where the compiler is not included e.g.

```toml
[external]
hdf5 = "1.10.1"
```
In the other case yes we could simply concatenate the key and value, but I am under the impression that it is bad practice to keep any configuration data within keys instead of values of the toml file. Which is why the current redundancy is preferable in my opinion. I am not married to it though. 

I am on gitter if you have more questions. I will be on irc later today.

**Comment by Spartee** [comment]
@ben-albrecht  @mppf Since we know can parse spec expressions with the addition of the SpecParser module, users can input the spec expression into the toml file in a number of ways. 

```toml
[external]
hdf5 = "hdf5@1.10.1%clang"
hdf5 = "1.10.1%clang"
hdf5 = "1.10.1" # Compiler inferred using CHPL_TARGET_COMPILER
hdf5 = "1.10.1%clang +hl+mpi" 
```

We could not do this before because we chose to let Spack completely handle the spec expression's but now with the SpecParser we have more control over what happens in the Spack installs and builds. 

The only caveat to this approach is the variants. For example if a user has two different variants of hdf5(same version and compiler e.g. `hdf5@1.10.1%clang`) but one uses the high level interface(`hdf5@1.10.1%clang +hl`) and one does not(`hdf5@1.10.1%clang ~hl`), currently we cannot decipher between them. The package that is found first in that search will be taken. Deciphering between them would mean more work on the parser to be able to compute which package hash is the correct version to take. This is not something that alot of users will run into unless they need multiple different variants of the same package with the same version and compiler on one machine. But I wanted to make a note that this is a problem that should be alleviated at some point in the future.

**Comment by mppf** [comment]
> Deciphering between them would mean more work on the parser to be able to compute which package hash is the correct version to take. 

I'm not so sure I understand. I was supposing that the purpose of SpecParser would be to create canonical Spack specs (e.g. adding target compiler if missing). So I'd suppose that it would leave variants alone, and then Spack would handle these.

Put another way, is supporting variants better a small or a big effort to fix?

**Comment by Spartee** [comment]
Currently it does leave the variants alone, you are correct. Maybe it would help to mention that we could not do this before either.

Supporting variants would probably require me talking to the people at Spack and possibly getting them to develop a better way of retrieving dependencies. With my current/knowledge of Spack and the commands I don't believe there is a better way although I am actively looking.

**Comment by Spartee** [comment]
Here's an example of where this problem might come into play.

```toml
[external]
hdf5 = "1.10.1"
```

but... this user has two of the same version of `HDF5` installed and built with the same compiler. The only difference is that one of the packages INCLUDES the high level interface variant and the other package EXCLUDES the high level interface variant. So say the user wanted to fix the problem by specifying that they wanted the `HDF5` package with the high level interface by saying this in their `Mason.toml` like so:

```toml
[external]
hdf5 = "1.10.1 +hl"
```

In a perfect world, where Spack gave us a better way of finding dependencies, this would work and the user would get the right package. However.. we cannot do this yet. So Mason will just grab the first `HDF5` package, version 1.10.1, built with CHPL_TARGET_COMPILER, that it sees. 

bottom line..
- users can use multiple versions of a package(not in the same project at once though of course)
- users can use packages of the same version but built with multiple compilers
- users will run into problems when using a package of which there are two copies downloaded of the same version, built with the same compiler, and only differ in variant. Mason, as of today, can only guess which one the user wants.

It's not a big deal for our first interface with Spack. It was more just a note. Hope this helps!

**Comment by mppf** [comment]
@Spartee - thanks for explaining. I'd still like to understand what is hard about it though. I would have expected that with this example:
```
[external]
hdf5 = "1.10.1 +hl"
```

when we ask for details about the HDF5 dependency from Spack, we run some Spack command and pass the entire spec (i.e. `hdf5@1.10.1 +hl`) and then Spack tells us about the appropriate libraries/flags/dependencies. It sounds (from your description) that Spack will work with the complete spec if/when we request installation of the package, but not to gather dependencies/libraries/flags? Isn't there just some Spack command we can run (conceptually like the pkg-config commands) to get this information? Why wouldn't we pass the whole spec to that command?

**Comment by Spartee** [comment]
You are right on. If a user passed a completely "concretized"(as spec puts it) spec then we would be able to find the correct package based on variant. When I was testing, however, I realized that most of the use cases involved users only putting partial variants. For example:

```
[external]
# How im assuming most users would use variant spec functionality
hdf5 = "1.10.1 +hl"
```

```
[external]
# What users would have to do in order for us to find package by variant today (concretized)
hdf5 = "hdf5@1.10.1%clang@9.0.0-apple~cxx~debug~fortran+hl+mpi+pic+shared~szip~threadsafe arch=darwin-sierra-x86_64"
```

As you can see there is a dearth of necessary information in the first example which inhibits our ability to find packages by variant. When installing Spack can guess these gaps of information for you. Right now, the only thing we can guess for the user is what compiler they want to use based on CHPL_TARGET_COMPILER. 

Its the "guessing what package the user wants" portion that is difficult. Spack solved this problem by concretizing all specs(even partial ones) that are listed for a build. We don't have this functionality, however, I have been looking at some ways of possibly doing so. Spack's outputs are not very machine readable/friendly which also makes this more difficult.
